File: lib/dither/facts.rb
# frozen_string_literal: true

module Dither
  # What an extension can be asked about, and how to compare it.
  #
  # A fact is a declared name, a type, and a path into the extension's own
  # fetched data. That is the whole contract: an extension that publishes
  # "next_meeting_in" gets a trigger for it, and the rule engine never learns
  # what a meeting is.
  module Facts
    TYPES = %w[duration number text boolean].freeze

    # Operators are declared per type so the editor can offer only the ones
    # that make sense. Asking whether a duration "contains" something is not a
    # question, and offering it invites rules that can never be true.
    Operator = Data.define :id, :label, :arity do
      def value? = arity == 1
    end

    OPERATORS = [
      Operator[id: "lt", label: "is less than", arity: 1],
      Operator[id: "lte", label: "is at most", arity: 1],
      Operator[id: "gt", label: "is more than", arity: 1],
      Operator[id: "gte", label: "is at least", arity: 1],
      Operator[id: "eq", label: "is", arity: 1],
      Operator[id: "neq", label: "is not", arity: 1],
      Operator[id: "contains", label: "contains", arity: 1],
      Operator[id: "present", label: "has any value", arity: 0],
      Operator[id: "absent", label: "is empty", arity: 0]
    ].freeze

    OPERATORS_BY_ID = OPERATORS.to_h { [it.id, it] }.freeze

    OPERATORS_FOR = {
      "duration" => %w[lt lte gt gte present absent],
      "number" => %w[lt lte gt gte eq neq present absent],
      "text" => %w[eq neq contains present absent],
      "boolean" => %w[eq present absent]
    }.freeze

    # One declared fact.
    Fact = Data.define :key, :label, :type, :path, :unit do
      def self.from attributes
        values = Hash(attributes).transform_keys(&:to_s)

        new(
          key: values["key"].to_s,
          label: values["label"] || values["key"].to_s.tr("_", " ").capitalize,
          type: TYPES.include?(values["type"]) ? values["type"] : "text",
          path: values["path"].to_s,
          unit: values["unit"]
        )
      end

      def operators = OPERATORS_FOR.fetch(type, []).filter_map { Facts.operator it }

      def valid? = !key.empty? && !path.empty?
    end

    def self.operator(id) = OPERATORS_BY_ID.fetch id.to_s, nil

    def self.declared(extension) = Array(extension.facts).map { Fact.from it }.select(&:valid?)

    def self.find extension, key
      declared(extension).find { it.key == key.to_s }
    end

    # Digs a fact out of whatever the extension last fetched. A path is dotted,
    # and a missing step answers nil rather than raising: an extension that has
    # not fetched yet has no facts, which is a real state and not an error.
    # A path is dotted, and a numeric step indexes an array: the first
    # departure is "transit.departures.0.delay". A missing step answers nil
    # rather than raising, because an extension that has not fetched yet has no
    # facts, which is a real state and not an error.
    def self.value data, path
      path.to_s.split(".").reduce(data) do |current, step|
        case current
          when Hash then current[step]
          when Array then step.match?(/\A\d+\z/) ? current[step.to_i] : nil
          else break nil
        end
      end
    end

    # rubocop:todo-next Metrics/CyclomaticComplexity
    def self.compare actual, operator_id, expected
      case operator_id.to_s
        when "present" then !blank?(actual)
        when "absent" then blank?(actual)
        when "contains" then actual.to_s.downcase.include? expected.to_s.downcase
        when "eq" then actual.to_s.casecmp?(expected.to_s)
        when "neq" then !actual.to_s.casecmp?(expected.to_s)
        else numeric_compare actual, operator_id, expected
      end
    end

    def self.numeric_compare actual, operator_id, expected
      return false if blank? actual

      left = Float actual, exception: false
      right = Float expected, exception: false

      return false unless left && right

      case operator_id.to_s
        when "lt" then left < right
        when "lte" then left <= right
        when "gt" then left > right
        when "gte" then left >= right
        else false
      end
    end

    def self.blank?(value) = value.nil? || (value.respond_to?(:empty?) && value.empty?)

    private_class_method :numeric_compare
  end
end

File: web/src/lib/facts.ts
import type { FactType } from "@/lib/extensions/manifest";

/**
 * Facts are how an extension makes itself available to triggers.
 *
 * The flow engine knows nothing about trains, weather or calendars. It knows
 * that a widget publishes named, typed values, and that you can compare them.
 * A calendar connector that declares `next_meeting_in` gets "show the commute
 * when a meeting starts within thirty minutes" for free, without the engine
 * learning what a meeting is.
 *
 * The declared type decides which comparisons are offered, so the editor
 * cannot build a rule that asks whether a duration "contains" something - a
 * condition that could never be true is not constructible.
 */

export interface Operator {
  id: string;
  label: string;
  /** Whether the editor should ask for a value to compare against. */
  needsValue: boolean;
}

export const OPERATORS: Record<string, Operator> = {
  lt: { id: "lt", label: "is less than", needsValue: true },
  lte: { id: "lte", label: "is at most", needsValue: true },
  gt: { id: "gt", label: "is more than", needsValue: true },
  gte: { id: "gte", label: "is at least", needsValue: true },
  eq: { id: "eq", label: "is", needsValue: true },
  neq: { id: "neq", label: "is not", needsValue: true },
  contains: { id: "contains", label: "contains", needsValue: true },
  present: { id: "present", label: "has any value", needsValue: false },
  absent: { id: "absent", label: "is empty", needsValue: false },
};

const OPERATORS_FOR: Record<FactType, string[]> = {
  duration: ["lt", "lte", "gt", "gte", "present", "absent"],
  number: ["lt", "lte", "gt", "gte", "eq", "neq", "present", "absent"],
  text: ["eq", "neq", "contains", "present", "absent"],
  boolean: ["eq", "present", "absent"],
};

export function operatorsFor(type: FactType): Operator[] {
  return (OPERATORS_FOR[type] ?? []).map((id) => OPERATORS[id]);
}

/**
 * Resolve a dotted path against fetched data. A numeric step indexes an array,
 * so `source_1.departures.0.minutes_until` reads the next train.
 */
export function valueAt(data: unknown, path: string): unknown {
  return path.split(".").reduce<unknown>((current, step) => {
    if (current === null || current === undefined) return undefined;
    if (Array.isArray(current)) return /^\d+$/.test(step) ? current[Number(step)] : undefined;
    if (typeof current === "object") return (current as Record<string, unknown>)[step];
    return undefined;
  }, data);
}

const isBlank = (value: unknown) =>
  value === null || value === undefined || value === "" || (Array.isArray(value) && !value.length);

function asNumber(value: unknown): number | undefined {
  if (typeof value === "number") return Number.isFinite(value) ? value : undefined;
  if (typeof value === "string" && value.trim() !== "") {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : undefined;
  }
  return undefined;
}

function asBoolean(value: unknown): boolean {
  if (typeof value === "boolean") return value;
  if (typeof value === "string") return value === "true" || value === "1" || value === "yes";
  return Boolean(value);
}

/**
 * Compare a fact's current value against what a condition asks for.
 *
 * Missing data answers false rather than throwing. A display whose weather
 * provider is down should keep showing the previous screen, not stop.
 */
export function compare(actual: unknown, operator: string, expected: unknown): boolean {
  switch (operator) {
    case "present":
      return !isBlank(actual);
    case "absent":
      return isBlank(actual);
    default:
      break;
  }

  if (isBlank(actual)) return false;

  switch (operator) {
    case "lt":
    case "lte":
    case "gt":
    case "gte": {
      const left = asNumber(actual);
      const right = asNumber(expected);
      if (left === undefined || right === undefined) return false;
      if (operator === "lt") return left < right;
      if (operator === "lte") return left <= right;
      if (operator === "gt") return left > right;
      return left >= right;
    }
    case "eq":
    case "neq": {
      const left = asNumber(actual);
      const right = asNumber(expected);
      const same =
        left !== undefined && right !== undefined
          ? left === right
          : typeof actual === "boolean" || typeof expected === "boolean"
            ? asBoolean(actual) === asBoolean(expected)
            : String(actual).toLowerCase() === String(expected).toLowerCase();
      return operator === "eq" ? same : !same;
    }
    case "contains":
      return String(actual).toLowerCase().includes(String(expected).toLowerCase());
    default:
      return false;
  }
}

/** How a condition reads back as a sentence, for the flow canvas and the trace. */
export function describe(factLabel: string, operator: string, expected: unknown): string {
  const op = OPERATORS[operator];
  if (!op) return factLabel;
  return op.needsValue ? `${factLabel} ${op.label} ${expected}` : `${factLabel} ${op.label}`;
}
